ians_controller/ians_controller.ino

```diff
@@ -55,88 +54,78 @@ IMU robot_imu;
-  
-  // ROS Node Setup
-  nh.initNode();
-  nh.advertise(lwheel);
-  nh.advertise(rwheel);
-//  nh.advertise(imu_data);
-  nh.advertise(quatz);
-  nh.advertise(quatw);
-//  nh.advertise(compass);
-  nh.subscribe(lmotor_sub);
-  nh.subscribe(rmotor_sub);
-  nh.subscribe(lwheel_vtarget_sub);
-  nh.subscribe(rwheel_vtarget_sub);
-  nh.subscribe(reset_encoder_sub);
+    Serial.begin(BAUD_RATE);
+    // Interrupt set up
+    encoder_timer.begin(ROS_publisher, ENCODER_RATE * 1000); // Convert to microseconds
+    PID_timer.begin(run_PID, PID_RATE * 1000); // Convert to microseconds
+
+    // ROS Node Setup
+    nh.initNode();
+    nh.advertise(lwheel);
+    nh.advertise(rwheel);
+    nh.advertise(quatz);
+    nh.advertise(quatw);
+    nh.subscribe(lwheel_vtarget_sub);
+    nh.subscribe(rwheel_vtarget_sub);
+    nh.subscribe(reset_encoder_sub);
 }
 
 // Main Loop
 void loop() {
-  // Sit and spin and wait for message publications from the Pi
-  nh.spinOnce();
+    // Sit and spin and wait for message publications from the Pi
+    nh.spinOnce();
 }
 
-void run_PID() {  
-  r_pid.pid_spin();
-  l_pid.pid_spin();
+void run_PID() {
+    r_pid.pid_spin();
+    l_pid.pid_spin();
 }
 
 void ROS_publisher() {
-  static double then;
-  
-  // Send the odom to the Pi for the nav stack
-  lwheel_msg.data = (lmotor_encoder.read() / 4);
-  rwheel_msg.data = (rmotor_encoder.read() / 4);
-
-  static int encoder_counter = 0;
-  encoder_counter++;
-  if (encoder_counter > 14) {
-    lwheel.publish(&lwheel_msg);
-    rwheel.publish(&rwheel_msg);
-    encoder_counter = 0;
-    // Publish IMU data
-    auto quat_data = robot_imu.read_IMUmsg_data();
-    quatw_msg = std::get<0>(quat_data);
-    quatz_msg = std::get<1>(quat_data);
-    quatw.publish(&quatw_msg);
-    quatz.publish(&quatz_msg);    
-//    imu_data.publish(&IMU_msg);
-  }
-  
-  // Update the PID controller with the current odom
-  
-  l_pid.cumulative_enc_val(lwheel_msg.data);
-  r_pid.cumulative_enc_val(rwheel_msg.data);
+<<<<<<< HEAD
+    // Send the odom to the Pi for the nav stack
+    lwheel_msg.data = (lmotor_encoder.read() / 4);
+    rwheel_msg.data = (rmotor_encoder.read() / 4);
+
+    // check for motor stall
+    l_halt_highlevel = l_pid.check_motor_stall(lwheel_msg.data);
+    r_halt_highlevel = r_pid.check_motor_stall(rwheel_msg.data);
+
+    static int encoder_counter = 0;
+    encoder_counter++;
+    if (encoder_counter > 14) {
+        lwheel.publish(&lwheel_msg);
+        rwheel.publish(&rwheel_msg);
+        encoder_counter = 0;
+        // Publish IMU data
+        auto quat_data = robot_imu.read_IMUmsg_data();
+        quatw_msg = std::get<0>(quat_data);
+        quatz_msg = std::get<1>(quat_data);
+        quatw.publish(&quatw_msg);
+        quatz.publish(&quatz_msg);
+    }
+
+    // Update the PID controller with the current odom
+    l_pid.cumulative_enc_val(lwheel_msg.data);
+    r_pid.cumulative_enc_val(rwheel_msg.data);
 }
 
 void lmotor_callback(const std_msgs::Float32& msg) {
```

Comment:
what does this function do? is it necessary? if so leave it, if not necessary then delete

Reply:
What function, `lmotor_callback`?